File: cpp/acwing/sum_of_submatrix.cpp
//
// Created by bruce on 2021/6/18.
//

/**
 * AcWing-796. 子矩阵的和
 */

#include <iostream>

using namespace std;

const int N = 1000 + 10;

int n, m, q;
int a[N][N], s[N][N];

int main() {
    // 第一行包含三个整数 n，m，q。
    scanf("%d%d%d", &n, &m, &q);
    // 接下来 n 行，每行包含 m 个整数，表示整数矩阵。
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            scanf("%d", &a[i][j]);
        }
    }

    // 二维前缀和
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            s[i][j] = s[i - 1][j] + s[i][j - 1] - s[i - 1][j - 1] + a[i][j];
        }
    }

    // 接下来 q 行，每行包含四个整数 x1,y1,x2,y2，表示一组询问。
    while (q--) {
        int x1, y1, x2, y2;
        scanf("%d%d%d%d", &x1, &y1, &x2, &y2);
        printf("%d\n", s[x2][y2] - s[x2][y1 - 1] - s[x1 - 1][y2] + s[x1 - 1][y1 - 1]);
    }

    return 0;
}

File: js/acwing/sum_of_submatrix.js
/**
 * @Description
 * AcWing-796. 子矩阵的和
 * 题目链接：https://www.acwing.com/problem/content/798/
 * @Author PAN Bruce
 * @Date 2021/6/18
 */

var fs = require('fs');
var buf = '';

process.stdin.on("readable", function () {
    var chunk = process.stdin.read();
    if (chunk) buf += chunk.toString();
});

process.stdin.on("end", function () {
    const lines = buf.split("\n");
    // 第一行包含三个整数 n，m，q
    let tokens = lines[0].split(" ").filter(s => s !== "").map(val => +val);
    const n = tokens[0],
        m = tokens[1],
        q = tokens[2];

    // 接下来 [1, n] 行，每行包含 m 个整数，表示整数矩阵。
    const arr = [];

    for (let i = 0; i < n; i++) {
        arr.push(lines[1 + i].split(" ").filter(s => s !== "").map(val => +val));
    }

    // 计算前缀矩阵
    const sum = [];
    for (let i = 0; i <= n; i++) {
        sum.push(new Array(m + 1).fill(0));
    }

    for (let i = 0; i < n; i++) {
        // sum[i + 1][j + 1] 对应 arr[0, 0] 到 arr[i][j]
        for (let j = 0; j < m; j++) {
            sum[i + 1][j + 1] = sum[i + 1][j] + sum[i][j + 1] - sum[i][j] + arr[i][j];
        }
    }

    // 接下来 q 行，每行包含四个整数 x1,y1,x2,y2，表示一组询问。
    // [n + 1, n + q]
    for (let i = 0; i < q; i++) {
        tokens = lines[n + i + 1].split(" ").filter(s => s !== "").map(val => +val);
        let [x1, y1, x2, y2] = tokens;
        // [x, y] -> 对应下标为 [x - 1, y - 1] -> sum[l][r]
        console.log(sum[x2][y2] - sum[x1 - 1][y2] - sum[x2][y1 - 1] + sum[x1 - 1][y1 - 1]);
    }
});

